src/utils/flagging.R

```diff
@@ -0,0 +1,306 @@
+#'     values. For instance, `180` would indicate flagging whenever there is a
+#'     positive value after 180 days.
+#' @param periods Numeric vector of times that flags are calculated, used
+#'     for all of the `thresholds_...` arguments. For instance, `c(7, 30, 90)`
+#'     would specify that flags are calculated for rolling sums of 7 days,
+#'     30 days, and 90 days.
+#' @param thresholds_pcts Numeric vector of percentile to use for flagging,
+#'     so that any time the value for `x` is over that percent quantile based
+#'     on the historic data, a flag is
+#'     generated. If a single value, is recycled, otherwise must have the
+#'     same length as `periods`.
+#' @param thresholds_static Numeric vector of static thresholds to use for
+#'     flagging, so that any time the value of `x` is above that value, a flag
+#'     is generated. If a single value, is recycled, otherwise must have the
+#'     same length as `periods`.
+#' @param thresholds_minimums Numeric vector of minimum values such that if `x`
+#'     is above it, no flags are generated. If a single value, is recyled,
+#'     otherwise must have the same length as `periods`.
+#'
+#' @return Data frame with rolling sums and flags generated.
+calculate_flags <- function(
+    .data,
+    date,
+    x,
+    first_since = NULL,
+    periods = NULL,
+    thresholds_pcts = NULL,
+    thresholds_static = NULL,
+    thresholds_minimums = NULL
+) {
+  # at least one of first_since and periods must be passed
+  null_p <- is.null(periods)
+  null_fs <- is.null(first_since)
+
+  if (null_p && null_fs) {
+    stop(
+      "At least one of `first_since` and `periods` must be passed.",
+      call. = FALSE
+    )
+  }
+
+  # if periods specified, check that all arguments match as necessary
+
+  if (!null_p) {
+    # we order all of the arguments by periods
+    # for later pmap
+    p_order <- order(periods, decreasing = TRUE)
+    periods <- periods[p_order]
+
+    null_tp <- is.null(thresholds_pcts)
+    null_ts <- is.null(thresholds_static)
+
+    if (null_tp && null_ts) {
+      stop(
+        "One of `thresholds_pcts` or `thresholds_static` must be ",
+        "passed to `calculate_alerts()` if `periods` is ",
+        "not NULL.",
+        call. = FALSE
+      )
+    }
+
+    # create thresholds that can't be met if null is set
+    if (null_tp) {
+      thresholds_pcts <- 1
+    }
+
+    if (null_ts) {
+      thresholds_static <- Inf
+    }
+
+    # check the lengths of the inputs
+    ln_p <- length(periods)
+    ln_tp <- length(thresholds_pcts)
+    ln_ts <- length(thresholds_static)
+
+
+    if ((ln_p != ln_tp && ln_tp != 1) || (ln_p != ln_ts && ln_ts != 1)) {
+      stop(
+        "`thresholds_pcts` and `thresholds_static` must have the same length ",
+        "as `periods` if passed to `calculate_alerts()`, unless length 1, in ",
+        " which case they are recycled.",
+        call. = FALSE
+      )
+    }
+
+    # reorderig for later pmap
+    if (ln_tp > 1) {
```

Comment:
So, you're right in that they have to be passed in the same order. So if you passed the vector you mention, you would get some strange results. But there isn't really a way we can check easily I think. Or so I thought. What I can do is generate a warning if the ordering isn't the same. So we expect the period length, as it goes up, to see monotonically increasing values of the thresholds, and will generate a warning if not. And we expect the same for the minimum values, to monotonically increase.

We won't stop the function with an error, since there could be edge cases, but good catch. I have also adjusted the documentation to meet this. Note that I am not going to check `thresholds_pct`, because I don't think we can expect the same behavior, you might want 90% for lower time spans and 95% for longer, but you might also want the opposite.

Reply:
Okay these functions implemented!